### PR TITLE
fix(frontend): fix picasso drag-and-drop issue with 0 length placeholder

### LINKED ISSUE #3463: Drag and drop bug on picasso
When dragging new schedule entries onto the picasso, under some yet unknown conditions, sometimes the release is not triggered, and the gray placeholder schedule entry gets stuck to the mouse pointer permanently:

[Bildschirmaufzeichnung vom 2023-05-15, 22-30-03 (getrimmt).webm](https://github.com/ecamp/ecamp3/assets/7566995/4b6bef4a-af7a-492a-867f-78067ddbedf9)

The bug has several, progressively worse stages. Timeline of the above video:
0:03 - 0:11 stage 1
0:12 - 0:32 stage 2
0:33 - 0:44 back in stage 1
0:44 user refreshes the page to get rid of the bug
(Stage 3 is never reached in this video, see below for a video of that)

### Stage 1: Canceling the dialog
I was able to reproduce this by double-clicking on the draggable picasso surface, while moving the mouse pointer enough in vertical direction between the two clicks. It has to be at least enough movement such that a minimal length schedule entry is created (i.e. **in very short succession: click, drag a little bit, click**):
[picasso-dragging-bug-1.webm](https://github.com/ecamp/ecamp3/assets/7566995/88c968b6-f131-4eaa-9171-8a223ac9f092)

### Stage 2: Getting the schedule entry end stuck to the cursor
Once a user is in stage 1, they might try to get rid of the gray placeholder and/or re-trigger the dialog, by dragging the length handle of the placeholder. This makes matters worse, because now, the end of the schedule entry is stuck to the mouse cursor, until the user clicks again, getting back to stage 1.
In some cases, we can also get directly to stage 2, when the second click of the reproduction above accidentally grabs and moves the length handle of the placeholder schedule entry.
[picasso-dragging-bug-2.webm](https://github.com/ecamp/ecamp3/assets/7566995/b669daa0-e703-4a5c-8553-8370d8b1e350)

### Stage 3: Getting the entire schedule entry stuck to the cursor
While in stage 1, if the user tries to move the placeholder schedule entry by dragging its body, the placeholder gets stuck permanently to the mouse cursor. I have found no way to get out of this situation, except for refreshing the page. In this stage, if the user tries to save the schedule entry, we get a console error `g._meta is undefined` or `Cannot read properties of undefined (reading 'self')` or `undefined is not an object (evaluating 'g._meta.self')` in the ScheduleEntries.vue component. These are collected in a [Sentry issue](https://ecamp.sentry.io/issues/4004671918/?project=5602507).
[picasso-dragging-bug-3.webm](https://github.com/ecamp/ecamp3/assets/7566995/59648ba8-4b8a-46ec-b019-b3b9eceabc06)

### PR DESCRIPTION
**Pull request by usu**
Fixes #3463 (hopefully)

also adresses https://github.com/ecamp/ecamp3/pull/3410#pullrequestreview-1384144228

Based on #3411 because it makes debugging much easier when seeing the start/end times directly. Happy to rebase on devel, in case we are not yet ready to merge #3411

I tried to only touch code directly related to this bug. Can you guys try it out, if the behaviour matches our expectations?

If so, I can check, if we have similar issues with the other drag & drop handlers as well (resize, move). Plus also cleaning up the code a bit.

### CODE REVIEW

File: common/helpers/dateHelperUTCFormatted.js [hunk 1]
```diff
@@ -37,6 +37,12 @@ function timeDurationShort(start, end, tc) {
   })
 }
 
+// short format of dateTime range
```
Comment (BacLuc) [inline common/helpers/dateHelperUTCFormatted.js:40]: Here you could use jsdoc, this would appears in the ide when you hover over the function or in the autocompletion when you use the function

**Comment by carlobeltrame** [comment]
Great, on the deployment I can't reproduce the bug anymore with my method!
Does this PR then also supersede #3410? The dragging in both directions feels much more natural to me now.
I think I'd like to rebase this on devel, because I'm not really happy with the details of #3411 yet, and I have a feeling we need to talk some details of that PR over.
I haven't understood all of your changes yet, but will do an in-depth review soon.

File: frontend/src/helpers/vCalendarDragAndDrop.js [hunk 2]
```diff
@@ -2,11 +2,22 @@
  * helpers for VCalendar DragAndDrop composables
  */
 
+const ONE_MINUTE = 60 * 1000
+const ONE_HOUR = 60 * ONE_MINUTE
+const ONE_DAY = 24 * ONE_HOUR
+
+const roundToMinutes = 15 // minutes
```
Comment (carlobeltrame) [inline frontend/src/helpers/vCalendarDragAndDrop.js:9]: It seems a little odd to have this in a constant here. Wouldn't that be better as a parameter of roundTime with a default value? Or otherwise maybe rename roundTime to roundTimeToQuarterHour?

File: frontend/src/components/program/picasso/PicassoEntry.vue [hunk 3]
```diff
@@ -379,6 +380,8 @@ export default {
 }
 .e-picasso-entry--editable:hover,
 .e-picasso-entry--temporary {
+  cursor: default;
```
Comment (carlobeltrame) [inline frontend/src/components/program/picasso/PicassoEntry.vue:383]: Why was this added?

File: frontend/src/helpers/vCalendarDragAndDrop.js [hunk 4]
```diff
@@ -2,11 +2,22 @@
...
+const roundToMilliseconds = roundToMinutes * ONE_MINUTE
+
 // helper function to convert Vuetify day & time object into timestamp
 const toTime = (tms) => {
   return new Date(tms.year, tms.month - 1, tms.day, tms.hour, tms.minute).getTime()
 }
 
+const roundTime = (time) => {
```
Comment (carlobeltrame) [inline frontend/src/helpers/vCalendarDragAndDrop.js:17]: Could we use this function in useDragAndDropMove.js as well, and then remove roundTimeDown entirely?

**Review by carlobeltrame — APPROVED** [review]
Consider adding tests for roundTime, like the ones we have for roundTimeDown.

File: common/helpers/dateHelperUTCFormatted.js (hunk 1 above)
Comment (usu) [reply common/helpers/dateHelperUTCFormatted.js:40]: This code is part of #3411

File: frontend/src/helpers/vCalendarDragAndDrop.js (hunk 2 above)
Comment (usu) [reply frontend/src/helpers/vCalendarDragAndDrop.js:9]: Renamed in #3468

File: frontend/src/helpers/vCalendarDragAndDrop.js (hunk 4 above)
Comment (usu) [reply frontend/src/helpers/vCalendarDragAndDrop.js:17]: Fixed in #3468

File: frontend/src/components/program/picasso/PicassoEntry.vue (hunk 3 above)
Comment (usu) [reply frontend/src/components/program/picasso/PicassoEntry.vue:383]: For the temporary placeholder, I don't want to show the grab pointer (which is for move operation of existing events). This change makes the switching back and forth of different cursor animations a bit less flaky. But still remains a cosmetic issue to fix.

**Comment by usu** [comment]
Superseded by #3468
